types/types.go

```diff
@@ -520,91 +520,56 @@ func Compare(a, b Type) int {
 	case nil, Null, Boolean, Number, String:
 		return 0
 	case *Array:
-		arrA := a.(*Array)
-		arrB := b.(*Array)
-		if arrA.dynamic != nil && arrB.dynamic == nil {
-			return 1
-		} else if arrB.dynamic != nil && arrA.dynamic == nil {
-			return -1
-		}
-		if arrB.dynamic != nil && arrA.dynamic != nil {
-			if cmp := Compare(arrA.dynamic, arrB.dynamic); cmp != 0 {
-				return cmp
-			}
-		}
-		return typeSliceCompare(arrA.static, arrB.static)
+		return typeArrayCompare(a, b)
 	case *Object:
-		objA := a.(*Object)
-		objB := b.(*Object)
-		if objA.dynamic != nil && objB.dynamic == nil {
-			return 1
-		} else if objB.dynamic != nil && objA.dynamic == nil {
-			return -1
-		}
-		if objA.dynamic != nil && objB.dynamic != nil {
-			if cmp := Compare(objA.dynamic.Key, objB.dynamic.Key); cmp != 0 {
-				return cmp
-			}
-			if cmp := Compare(objA.dynamic.Value, objB.dynamic.Value); cmp != 0 {
-				return cmp
-			}
-		}
+		return typeObjectCompare(a, b)
+	case *Set:
+		return typeSetCompare(a, b)
+	case Any:
+		return typeAnyCompare(a, b)
+	case *Function:
+		return typeFunctionCompare(a, b)
+	default:
+		panic("unreachable")
+	}
+}
 
-		lenStaticA := len(objA.static)
-		lenStaticB := len(objB.static)
+// CompareTypeDiff compares the given inputs and returns the topmost
+// differing types between the inputs
+func CompareTypeDiff(a, b Type) ([]Type, []Type) {
```

Comment:
I'd prefer if the type eliding was an implementation detail of the error reporting in the type checker. If we see a need for this helper elsewhere, we could expose it--but in the spirit of keeping the package export surface area as small as possible, I'd like to see this made private. That means it'll have to go into the AST package.

Reply:
Same comment applies below for the formatting.